File: scala/src/main/scala/clase1/PatternMatching_Max.scala
package clase1

object PatternMatching_Max extends App {
  val max = (list: List[Int]) => {

    def doMax(list: List[Int], acc: Int): Int = list match {
      case head :: tail => doMax(tail, if(head > acc) head else acc)
      case Nil => acc
    }

    doMax(list, Int.MinValue)

  }

  println(max(List(12, 3, 45, 22, 115, 4)))
  println(max(List()))

}

File: haskell/src/clase1/PatternMatchingMax.hs
module PatternMatchingMax where

doMax (head:tail) acc = doMax tail (if head > acc then head else acc)
doMax [] acc = acc

maxList list = doMax list minBound :: Int

main = do
    print (maxList [12, 3, 45, 22, 115, 4])
    print (maxList [])